Sources/FeaturevisorSDK/Instance.swift

```diff
@@ -251,38 +282,193 @@ public class FeaturevisorInstance {
+                        enabled: requiredFeaturesAreEnabled)
+
+                return evaluation;
+            }
+        }
+
+        // bucketing
+        let bucketValue = getBucketValue(feature: feature, context: finalContext);
+
+        let matchedTraffic = getMatchedTraffic(
+                traffic: feature.traffic,
+                context: finalContext,
+                datafileReader: datafileReader)
+
+        if let matchedTraffic {
+
+            if !feature.ranges.isEmpty {
+
+                let matchedRange = feature.ranges.first(where: { range in
+                    return bucketValue >= range.start && bucketValue < range.end
+                });
+
+                // matched
+                if (matchedRange != nil) {
+                    evaluation = Evaluation(
+                            featureKey: feature.key,
+                            reason: .allocated,
+                            bucketValue: bucketValue,
+                            enabled: matchedTraffic.enabled ?? true)
+
+                    return evaluation;
+                }
+
+                // no match
+                evaluation = Evaluation(
+                        featureKey: feature.key,
+                        reason: .outOfRange,
+                        bucketValue: bucketValue,
+                        enabled: false)
+
+                logger.debug("not matched", ["featureKey": featureKey]) // TODO: Log evaluation object. Make it encodable
+
+                return evaluation;
+            }
+
+            // override from rule
+            if let matchedTrafficEnabled = matchedTraffic.enabled {
+                evaluation = Evaluation(
+                        featureKey: feature.key,
+                        reason: .override,
+                        bucketValue: bucketValue,
+                        ruleKey: matchedTraffic.key,
+                        enabled: matchedTrafficEnabled,
+                        traffic: matchedTraffic)
+
+                logger.debug("override from rule", ["featureKey": featureKey]) // TODO: Log evaluation object. Make it encodable
+
+                return evaluation;
+            }
+
+            // treated as enabled because of matched traffic
+            if bucketValue < matchedTraffic.percentage {
+                // @TODO: verify if range check should be inclusive or not
+                evaluation = Evaluation(
+                        featureKey: feature.key,
+                        reason: .rule,
+                        bucketValue: bucketValue,
+                        ruleKey: matchedTraffic.key,
+                        enabled: true,
+                        traffic: matchedTraffic)
+
+                return evaluation
+            }
+        }
+
+        // nothing matched
+        evaluation = Evaluation(
+            featureKey: feature.key,
+            reason: .error,
+            bucketValue: bucketValue,
+            enabled: false)
+
+        return evaluation;
+    }
```

Comment:
it's all coming together so beautifully! thanks @polok! 😭

Reply:
Hopefully, I won't find too much bugs when come to testing the implementation ;)